pkg/controller/pipelinerun/controller.go

```diff
@@ -0,0 +1,241 @@
+	pi "github.com/knative/build-pipeline/pkg/client/informers/externalversions/pipeline/v1alpha1"
+	listers "github.com/knative/build-pipeline/pkg/client/listers/pipeline/v1alpha1"
+	"github.com/knative/build-pipeline/pkg/constants"
+	"github.com/knative/build-pipeline/pkg/controller"
+	"github.com/knative/pkg/logging/logkey"
+	"go.uber.org/zap"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	typedcorev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/workqueue"
+)
+
+// for testing
+var (
+	runsListerFunc      = runsLister
+	pipelinesListerFunc = pipelinesLister
+	getRecordFunc       = getRecorder
+)
+
+// Controller defines the Pipelinerun Controller which runs the pipleine.
+type Controller struct {
+	// kubeClientset is a standard kubernetes clientset
+	kubeClientset kubernetes.Interface
+	// clientset is a clientset for our own API group
+	clientset       clientset.Interface
+	runsLister      listers.PipelineRunLister
+	pipelinesLister listers.PipelineLister
+	recorder        record.EventRecorder
+	logger          *zap.SugaredLogger
+	synced          cache.InformerSynced
+	workqueue       workqueue.RateLimitingInterface
+}
+
+// NewController returns a new pipeline run controller
+func NewController(
+	k kubernetes.Interface,
+	c clientset.Interface,
+	l *zap.SugaredLogger,
+) controller.Interface {
+
+	//TODO: Read timeout from a global config
+	pipelineFactory := informers.NewSharedInformerFactory(c, time.Second*30)
+	runsInformer := pipelineFactory.Pipeline().V1alpha1().PipelineRuns()
+	informer := pipelineFactory.Pipeline().V1alpha1().Pipelines()
+
+	// Enrich the logs with controller name
+	l = l.Named(constants.PipelineRunAgent).With(zap.String(logkey.ControllerType, constants.PipelineRunAgent))
+
+	controller := &Controller{
+		kubeClientset:   k,
+		clientset:       c,
+		runsLister:      runsListerFunc(runsInformer),
+		pipelinesLister: pipelinesListerFunc(informer),
+		synced:          runsInformer.Informer().HasSynced,
+		workqueue:       workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "PipelineRuns"),
+		recorder:        getRecordFunc(k, l),
+		logger:          l,
+	}
+
+	l.Info("Setting up event handlers")
+	// Set up an event handler for when PipelineRun resource is added
+	runsInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: controller.enqueuePipelineRun,
+	})
+	return controller
+}
+
+// Run will set up the event handlers for types we are interested in, as well
+// as syncing informer caches and starting workers. It will block until stopCh
+// is closed, at which point it will shutdown the workqueue and wait for
+// workers to finish processing their current work items.
```

Comment:
great comment! maybe include what `threadiness` is as well?

Reply:
This is removed now.